lib/src/config_document_parser.cc

```diff
@@ -0,0 +1,632 @@
+                } else if (!had_open_curly) {
+                    throw parse_exception(add_quote_suggestion(t->to_string(),
+                           "unbalanced close brace '}' with no open brace", _equals_count > 0));
+                }
+                object_nodes.push_back(make_shared<config_node_single_token>(tokens::close_curly_token()));
+                break;
+            } else if (t->get_token_type() == token_type::END && !had_open_curly) {
+                put_back(t);
+                break;
+            } else if (_flavor != config_syntax::JSON && is_include_keyword(t)) {
+                shared_node_list include_nodes;
+                include_nodes.push_back(make_shared<config_node_single_token>(t));
+                shared_ptr<config_node_include> inc = parse_include(include_nodes);
+                object_nodes.push_back(inc);
+                after_comma = false;
+            } else {
+                key_value_nodes.clear();
+                shared_token key_token = t;
+                auto key_path = parse_key(key_token);
+                key_value_nodes.push_back(key_path);
+                shared_token after_key = next_token_collecting_whitespace(key_value_nodes);
+                bool inside_equals = false;
+
+                shared_node_value next_value = nullptr;
+                if (_flavor == config_syntax::CONF && after_key->get_token_type() == token_type::OPEN_CURLY) {
+                    // can omit the ':' or '=' befor an object value
+                    next_value = parse_value(after_key);
+                } else {
+                    if (!is_key_value_separator(after_key)) {
+                        throw parse_exception(add_quote_suggestion(after_key->to_string(),
+                        "Key '" + key_path->render() + "' may not be followed by token: " + after_key->to_string(),
+                        _equals_count > 0));
+                    }
+
+                    key_value_nodes.push_back(make_shared<config_node_single_token>(after_key));
+                    if (after_key->get_token_type() == token_type::EQUALS) {
+                        inside_equals = true;
+                        _equals_count++;
+                    }
+
+                    next_value = consolidate_values(key_value_nodes);
+                    if (!next_value) {
+                        next_value = parse_value(next_token_collecting_whitespace(key_value_nodes));
+                    }
+                }
+
+                key_value_nodes.push_back(next_value);
+                if (inside_equals) {
+                    _equals_count--;
+                }
+                last_inside_equals = inside_equals;
+
+                string key = *key_path->get_path().first();
+                path remaining = key_path->get_path().remainder();
+
+                if (remaining.empty()) {
+                    try {
+                        keys.at(key);
```

Comment:
This is an expensive way to check for existence. There are two simpler ways, depending on what you need out of it
- Check for keys.find(key) == keys.end()
- Just insert/emplace, and check the return value. If the key already exists, the insert/emplace will not overwrite it.
http://www.cplusplus.com/reference/unordered_map/unordered_map/emplace/

So I'd rewrite this as
```
auto emplace = keys.emplace(move(key), true);
if (!emplace.second) {
    if (_flavor...) { ... }
    // This is a little unclear, but the first item is a pair<key,value> reference, so we can update the value by overwriting second.
    emplace.first.second = true;
}
```